src/nifreeze/data/filtering.py

```diff
@@ -96,3 +102,122 @@ def advanced_clip(
+
+    :math:`p_{5}` and :math:`p_{95}` being the 5th percentile and the 95th percentile of the data,
+    respectively.
+
+    If a mask is provided, only the data within the mask are considered.
+
+    Parameters
+    ----------
+    data : :obj:`~numpy.ndarray`
+        Data to be detrended.
+    mask : :obj:`~numpy.ndarray`, optional
+        Mask. If provided, only the data within the mask are considered.
+
+    Returns
+    -------
+    :obj:`~numpy.ndarray`
+        Detrended data.
+    """
+
+    data = data.copy().astype("float32")
+    reshaped_data = data.reshape((-1, data.shape[-1])) if mask is None else data[mask]
+    p5 = np.percentile(reshaped_data, 5.0, axis=0)
+    p95 = np.percentile(reshaped_data, 95.0, axis=0) - p5
+    return (data - p5) * p95.mean() / p95 + p5.mean()
+
+
+def detrend_dwi_median(data: np.ndarray, mask: np.ndarray | None = None) -> np.ndarray:
+    """Detrend DWI data.
+
+    Regresses out global DWI signal differences so that its standardized and centered around the
+    :data:`src.nifreeze.model.base.DEFAULT_CLIP_PERCENTILE` percentile.
+
+    If a mask is provided, only the data within the mask are considered.
+
+    Parameters
+    ----------
+    data : :obj:`~numpy.ndarray`
+        Data to be detrended.
+    mask : :obj:`~numpy.ndarray`, optional
+        Mask. If provided, only the data within the mask are considered.
+
+    Returns
+    -------
+    :obj:`~numpy.ndarray`
+        Detrended data.
+    """
+
+    shelldata = data[..., mask]
+
+    centers = np.median(shelldata, axis=(0, 1, 2))
+    reference = np.percentile(centers[centers >= 1.0], DEFAULT_CLIP_PERCENTILE)
+    centers[centers < 1.0] = reference
+    drift = reference / centers
+    return shelldata * drift
+
+
+def clip_dwi_shell_data(
+    data: np.ndarray,
+    gradients: np.ndarray,
+    index: int,
+    th_low: float = BVAL_THRESHOLD,
+    th_high: float = BVAL_THRESHOLD,
+) -> np.ndarray:
+    """Clip DWI shell data around the given index and lower and upper b-value bounds.
+
+    Clip DWI data around the given index with the provided lower and upper bound b-values.
+
+    Parameters
+    ----------
+    data : :obj:`~numpy.ndarray`
+        Data to be clipped.
+    gradients : :obj:`~numpy.ndarray`
+        Gradients.
+    index : :obj:`int`
+        Index of the shell data.
+    th_low : :obj:`float`, optional
+        A lower bound for the b-value.
+    th_high : :obj:`float`, optional
+        An upper bound for the b-value.
+
+    Returns
+    -------
+    clipped_dataset : :obj:`~numpy.ndarray`
+        Clipped data.
+    """
+
+    clipped_data = copy.deepcopy(data)
+
+    bvalues = gradients[:, -1]
+    bcenter = bvalues[index]
+
+    shellmask = np.ones(clipped_data.shape[-1], dtype=bool)
+
+    # Keep only bvalues within the range defined by th_high and th_low
+    shellmask[index] = False
+    shellmask[bvalues > (bcenter + th_high)] = False
+    shellmask[bvalues < (bcenter - th_low)] = False
```

Comment:
Return just shellmask?

Reply:
:+1: Agreed. Will do after committing the batch of suggestions.

Reply:
Done in 641067609fa91c0e77bceae5c3b30d1deac60bed.